### PR TITLE
Set maximum viewing range to 4000

### PR DESCRIPTION
**Pull request by Rogier-5**
Edit: changed PR: limit viewing range to 4000 max, as that is about the maximum the network protocol supports.

------------------------------------------------------------------------

Using u8 will cause the wanted_range to wrap around if the player
selects a viewing range of 4096 or higher.

I realize this change makes yesterday's clients incompatible with
today's servers, and vice versa. However, people using development
versions might be expected to experience breakage now and then ?

This change is based on PR #4882.

### CODE REVIEW

**Comment by lhofhansl** [comment]
So this was a specific request from @paramat, @celeron55 and others. We a bunch of discussions around it.
👎 on warming that discussion again. 4096 nodes distance is far enough.

**Comment by Rogier-5** [comment]
> on warming that discussion again. 4096 nodes distance is far enough.

If u8 is deliberate and desired, that's fine with me. I don't specifically want it to be a u16 instead of u8.

I do want to fix the wraparound problem that happens when the player selects a viewing range of 4096. I'll recode.

**Comment by lhofhansl** [comment]
Who's doing a view range > 4096? That was something we deliberately accepted in #4811.

It's cool to guard against it it course, but looks a bit like a fix on search of a problem :)

**Comment by paramat** [comment]
:-1: u8 was chosen to make these extra packets as small as possible.
You can limit the setting when it is entered in advanced settings menu by adding the min/max value limitations in settingtypes.txt.

**Comment by Rogier-5** [comment]
Modified PR to impose a maximum on the viewing range.

Also modified settingtypes.txt and game.cpp to document and enforce the maximum.

**Comment by est31** [comment]
:+1: looks good in the current state (where it adds a maximum viewing range of 400 and a minimum of 20), and I think it can go in as bugfix.

File: src/client.cpp [hunk 1]
```diff
@@ -939,7 +940,9 @@ void writePlayerPos(LocalPlayer *myplayer, ClientMap *clientMap, NetworkPacket *
 	u32 keyPressed   = myplayer->keyPressed;
 	// scaled by 80, so that pi can fit into a u8
 	u8 fov           = clientMap->getCameraFov() * 80;
-	u8 wanted_range  = clientMap->getControl().wanted_range / MAP_BLOCKSIZE;
+	int wanted_range = std::ceil(clientMap->getControl().wanted_range / MAP_BLOCKSIZE);
```
Comment (rubenwardy) [inline src/client.cpp]: Isn't this an integer division?

```cpp
std::ceil((float)clientMap->getControl().wanted_range / (float)MAP_BLOCKSIZE);
```

File: src/game.cpp [hunk 2]
```diff
@@ -4214,7 +4224,7 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats,
 	if (draw_control->range_all) {
 		runData->fog_range = 100000 * BS;
 	} else {
-		runData->fog_range = 0.9 * draw_control->wanted_range * BS;
+		runData->fog_range = draw_control->wanted_range * BS;
```
Comment (rubenwardy) [inline src/game.cpp]: unrelated change

File: src/client.cpp (hunk 1 above)
Comment (Rogier-5) [inline src/client.cpp]: Actually, `clientMap->getControl().wanted_range` happens to be a float. I had also assumed is was integral, and my original code did and all-integer computation (see comments above), but it turned out to be a float, so I switched to `std::ceil()` instead.

Maybe I should make wanted_range integral, while I'm at it ?

File: src/game.cpp (hunk 2 above)
Comment (Rogier-5) [inline src/game.cpp]: While I do agree, some previous PRs which I had taken care to split into different commits were squashed on merge, so I'm a bit confused as to what is expected.

I'll be more than happy to split this PR into two commits.

Comment (rubenwardy) [inline src/game.cpp]: this change move the fog further out rather than at 90% (as discussed in other issues) - I don't see how this is related to this PR

File: src/client.cpp (hunk 1 above)
Comment (est31) [inline src/client.cpp]: No, its fine this way.

File: src/game.cpp (hunk 2 above)
Comment (lhofhansl) [inline src/game.cpp]: This is (and should be, IMHO) fixed in #4882.
@Rogier-5 are you planning on closing #4882 in favor of this?

**Comment by lhofhansl** [comment]
IMHO, #4882 should be merged first. Then this one should be rebased on top of that.

File: src/game.cpp (hunk 2 above)
Comment (Rogier-5) [inline src/game.cpp]: @lhofhansl I based this on #4882, as it modifies exactly the same line.

The intent was to rebase this once #4882 got merged. This is a different change after all.

Rebasing...

**Comment by Rogier-5** [comment]
Rebased.

File: src/client.cpp [hunk 3]
```diff
@@ -940,7 +940,9 @@ void writePlayerPos(LocalPlayer *myplayer, ClientMap *clientMap, NetworkPacket *
 	u32 keyPressed   = myplayer->keyPressed;
 	// scaled by 80, so that pi can fit into a u8
 	u8 fov           = clientMap->getCameraFov() * 80;
-	u8 wanted_range  = std::ceil(clientMap->getControl().wanted_range / MAP_BLOCKSIZE);
+	int wanted_range = std::ceil(clientMap->getControl().wanted_range / MAP_BLOCKSIZE);
```
Comment (lhofhansl) [inline src/client.cpp]: Use MYMAX here? (not important just a nit)

Comment (lhofhansl) [inline src/client.cpp]: Err... MYMIN I mean.

`u8 wanted_range = MYMIN(std::ceil(clientMap->getControl().wanted_range / MAP_BLOCKSIZE), 255);`

(again just a nit)

Comment (Rogier-5) [inline src/client.cpp]: Better indeed. Done.

File: src/game.cpp (hunk 2 above)
Comment (paramat) [inline src/game.cpp]: Yes this change belongs in #4882

**Comment by paramat** [comment]
Seems good now.

**Comment by paramat** [comment]
02112f84e2102182816ffa48d5444a7321730936
